src/main/java/io/fusionauth/http/io/DelegatingOutputStream.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2022, FusionAuth, All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+package io.fusionauth.http.io;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.GZIPOutputStream;
+
+import io.fusionauth.http.HTTPValues.ContentEncodings;
+import io.fusionauth.http.HTTPValues.Headers;
+import io.fusionauth.http.server.HTTPRequest;
+import io.fusionauth.http.server.HTTPResponse;
+
+/**
+ * A delegating output stream that can allow for latent configuration changes prior to writing to the underlying output stream.
+ *
+ * @author Daniel DeGroff
+ */
+public class DelegatingOutputStream extends OutputStream {
+  private final HTTPRequest request;
+
+  private final HTTPResponse response;
+
+  private final OutputStream unCompressingOutputStream;
+
+  private Boolean compress;
```

Comment:
This is a tri-state, so I may as well just make this an enum. `[unset, true, false]`.